### PR TITLE
Optimization/buildhist/colwisebuildhist

### PR DESCRIPTION
**Pull request by razdoburdin**
Hi,
this PR is a part of https://github.com/dmlc/xgboost/pull/7192. It continues the optimization of BuildHistKernel, that was started in https://github.com/dmlc/xgboost/pull/8218.
Here I introduce a column wise building of histogram and a dispatcher for choice between the row wise and the column wise kernels. This optimization allow to improve runtime speed for several datasets up to 2x. For performance measurements I booked the c6i.12xlarge (24 cores, hyperthreading is off) instance on on Amazon Web Services and used the benchmarks from [here](https://github.com/IntelPython/scikit-learn_bench).

Methodology of the benchmarking is the same as in https://github.com/dmlc/xgboost/pull/8218. The only two datasets from the list affecting by this optimization are santander and epsilon. For this reason only them are shown in the following table:
|                       | master    | this PR   | speedup |
| --------------------- | -------- | -------- | ------- |
| santander             | 1.02E+02 | 4.89E+01 | 2.09    |
| epsion                | 1.84E+02 | 1.03E+02 | 1.78    |

I am looking forward for your review and comments!

upg:
I have deleted the optimization related to santander dataset for simplification of the review.

### CODE REVIEW

**Comment by liuwenyi2** [comment]
Hi,
Your work is very interesting. I‘d like to test your optimization effect. Could you please tell me the source version of XGBoost?

**Comment by razdoburdin** [comment]
Hi,
for performance measurements I have used the master branch in the dmlc/xgboost repo.
You can easily reproduce the calculations with it.
If you need any help, you are welcome!

**Comment by liuwenyi2** [comment]
> Hi, for performance measurements I have used the master branch in the dmlc/xgboost repo. You can easily reproduce the calculations with it. If you need any help, you are welcome! From: wenyi Liu ***@***.***> Sent: Saturday, September 10, 2022 11:34 AM To: dmlc/xgboost ***@***.***> Cc: Razdoburdin, Dmitry ***@***.***>; Author ***@***.***> Subject: Re: [dmlc/xgboost] Optimization/buildhist/colwisebuildhist (PR #8233) Hi, Your work is very interesting. I‘d like to test your optimization effect. Could you please tell me the source version of XGBoost?

much appreciated!
Since I can only compile XGboost-1.1.0 on clusterIs.
I want to reproduce your calculations with 1.1.0 to do some measurements.
I'm trying to do it.

File: src/common/hist_util.cc [hunk 1]
```diff
@@ -204,75 +221,161 @@ void BuildHistKernel(const std::vector<GradientPair> &gpair,
...
+    ColsWiseBuildHistKernel<BinIdxType, first_page, any_missing>
+                           (gpair, row_indices, gmat, hist, hbm);
+  } else {
+    RowsWiseBuildHistKernel<do_prefetch, BinIdxType, first_page, any_missing>
+                           (gpair, row_indices, gmat, hist);
+  }
+}
+
+template <bool do_prefetch, bool first_page, bool any_missing, class GHistBuildingManager>
+void BuildHistDispatch(const std::vector<GradientPair> &gpair,
```
Comment (trivialfis) [inline src/common/hist_util.cc:286]: Can we use the `DispatchBinType` instead?

File: src/common/column_matrix.cc [hunk 2]
```diff
@@ -23,10 +23,12 @@ void ColumnMatrix::InitStorage(GHistIndexMatrix const& gmat, double sparse_thres
   gmat.GetFeatureCounts(feature_counts.data());
 
   // classify features
+  any_sparse_column_ = false;
```
Comment (trivialfis) [inline src/common/column_matrix.cc:26]: `any_sparse_column_ = !all_dense_column`

File: src/tree/hist/histogram.h [hunk 3]
```diff
@@ -16,6 +19,19 @@
 
 namespace xgboost {
 namespace tree {
+
+struct ColSample {
```
Comment (trivialfis) [inline src/tree/hist/histogram.h:23]: I think using `TrainParam` directly seems to be simpiler.

File: src/tree/hist/histogram.h [hunk 4]
```diff
@@ -26,9 +42,15 @@ class HistogramBuilder {
...
   BatchParam param_;
   int32_t n_threads_{-1};
+
+  ColSample colsample_;
+  std::shared_ptr<common::ColumnSampler> column_sampler_;
+
   size_t n_batches_{0};
   // Whether XGBoost is running in distributed environment.
   bool is_distributed_{false};
+  // Addhoch colsample threshold level
```
Comment (trivialfis) [inline src/tree/hist/histogram.h:52]: `adhoc`.

File: src/tree/hist/histogram.h [hunk 5]
```diff
@@ -59,10 +84,17 @@ class HistogramBuilder {
                             GHistIndexMatrix const &gidx,
                             std::vector<ExpandEntry> const &nodes_for_explicit_hist_build,
                             common::RowSetCollection const &row_set_collection,
-                            const std::vector<GradientPair> &gpair_h) {
+                            const std::vector<GradientPair> &gpair_h, int depth) {
     const size_t n_nodes = nodes_for_explicit_hist_build.size();
     CHECK_GT(n_nodes, 0);
 
+    const common::ColumnMatrix& column_matrix = gidx.Transpose();
+    const bool column_sampling =
```
Comment (trivialfis) [inline src/tree/hist/histogram.h:92]: Could you please explain the logic here? there are missing values but no sparse column in column matrix and bytree/bylevel lesser than threshold while there's no bynode? Is this necessary?

Also, this special case seems to be difficult to test.

**Review by trivialfis — COMMENTED** [review]
Thank you for the work on optimization. Some questions in the comments. Can we omit the optimization on column sampling for now and just build the histogram for all columns?

**Comment by razdoburdin** [comment]
> 
Hi, I have removed column sampling optimization for a while. Hope it will help with a review. Thanks for the suggestion!

File: src/common/hist_util.cc (hunk 1 above)
Comment (razdoburdin) [reply src/common/hist_util.cc:268]: done

**Review by trivialfis — COMMENTED** [review]
Could you please try to enable some tests for each of the case? I don't want to break anything in here by accident. ;-)

**Comment by razdoburdin** [comment]
> Could you please try to enable some tests for each of the case? I don't want to break anything in here by accident. ;-)

As far as column wise building is just another way to calculate the same result as the row wise building,  I suggest to reuse the existing tests for histogram building.
I added force_read_by_column flag to BuildHist. It allows to use disable automatic choice of building strategy and to use column wise building instead. I duplicated the related tests with the new flag, so they now test both strategies.

**Review by trivialfis — APPROVED** [review]

**Pull request merged**